llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -422,13 +422,16 @@ doPromotion(Function *F, FunctionAnalysisManager &FAM,
 
 /// Return true if we can prove that all callees pass in a valid pointer for the
 /// specified function argument.
-static bool allCallersPassValidPointerForArgument(Argument *Arg,
-                                                  Align NeededAlign,
-                                                  uint64_t NeededDerefBytes) {
+static bool allCallersPassValidPointerForArgument(
+    Argument *Arg, SmallPtrSet<CallBase *, 4> &RecursiveCalls,
+    Align NeededAlign, uint64_t NeededDerefBytes) {
   Function *Callee = Arg->getParent();
   const DataLayout &DL = Callee->getParent()->getDataLayout();
   APInt Bytes(64, NeededDerefBytes);
 
+  if (RecursiveCalls.size())
+    return true;
```

Comment:
Again, you can't just check that there's a recursive call; even if there's a recursive call, the checks in this function are relevant for the other calls.

Reply:
I don't think there is an easy way to implement this. I have pushed another patch but it fails even for the current testcase, but the workaround for it creates issues with other testcases too.

problem is that it looks at the recursive function's callee in the attached testcase and it fails at `isDereferenceableAndAlignedPointer` check for that call. So, I am not sure then if the attached testcase is correct in the first place ?

Reply:
Reading the testcase... the argument isn't marked dereferenceable for the definition of foo(), and we don't have any information in bar(), so the current logic can't prove it's dereferenceable.

It might be possible to prove the argument is dereferenceable because there's a load on all paths through foo(), but that logic shouldn't be in ArgPromotion; it should be in function-attrs, or something like that.

Reply:
> Reading the testcase... the argument isn't marked dereferenceable for the definition of foo(), and we don't have any information in bar(), so the current logic can't prove it's dereferenceable.

Yes, that's true. Do, you suggest I add the !dereferenceable metadata to this and run the testcase ?

> It might be possible to prove the argument is dereferenceable because there's a load on all paths through foo(), but that logic shouldn't be in ArgPromotion; it should be in function-attrs, or something like that.

Makes sense

Reply:
> Do, you suggest I add the !dereferenceable metadata to this and run the testcase ?

I was thinking a dereferenceable attribute, not metadata.  But yes, that probably makes sense.

> can this be a part of a separate patch

Sure.

Reply:
> > Do, you suggest I add the !dereferenceable metadata to this and run the testcase ?
> 
> I was thinking a dereferenceable attribute, not metadata. But yes, that probably makes sense.

```
define internal i32 @foo(ptr dereferenceable(16) %x, i32 %n, i32 %m) {
entry:
  %tmp = icmp ne i32 %n, 0
  br i1 %tmp, label %cond_true, label %cond_false

cond_true:                                        ; preds = %entry
  %tmp2 = load i32, ptr %x, align 4
  br label %return

cond_false:                                       ; preds = %entry
  %tmp5 = load i32, ptr %x, align 4
  %tmp7 = sub i32 %n, 1
  %tmp9 = call i32 @foo(ptr %x, i32 %tmp7, i32 %tmp5)
  %tmp11 = sub i32 %n, 2
  %tmp14 = call i32 @foo(ptr %x, i32 %tmp11, i32 %m)
  %tmp15 = add i32 %tmp9, %tmp14
  br label %return

cond_next:                                        ; No predecessors!
  br label %return

return:                                           ; preds = %cond_next, %cond_false, %cond_true
  %retval.0 = phi i32 [ %tmp2, %cond_true ], [ %tmp15, %cond_false ], [ undef, %cond_next ]
  ret i32 %retval.0
}

define i32 @bar(ptr %x, i32 %n, i32 %m) {
entry:
  %load0 = load i32, ptr %x, align 4
  %tmp3 = call i32 @foo(ptr %x, i32 %n, i32 %m)
  br label %return

return:                                           ; preds = %entry
  ret i32 %tmp3
}
``` 
I tried something like this, doesn't seem to work either! Can you take a look what might be wrong ?

> > can this be a part of a separate patch
> 
> Sure.

Reply:
The code currently only looks for dereferenceable attributes on the caller (i.e. bar()), not the function itself.

Reply:
> The code currently only looks for dereferenceable attributes on the caller (i.e. bar()), not the function itself.

Even this doesn't seem to work or manually adding a load i32 on ptr %x

```
define i32 @bar(ptr dereferenceable(32) %x, i32 %n, i32 %m) {
entry:
  %tmp3 = call i32 @foo(ptr %x, i32 %n, i32 %m)
  br label %return

return:                                           ; preds = %entry
  ret i32 %tmp3
}
```

Reply:
ping @efriedma-quic

Reply:
You might need to add an align attribute?  Otherwise, I'd suggest using a debugger to step into the dereferenceable check to see why it's failing.

Reply:
@efriedma-quic sorry for the tag, but I was able to get it working with align. Pushed a cleaned up patch, hope it is in a mergeable state now :) and can get a LGTM!